// @firestitch/datepicker/esm2015/libs/dialog/modules/desktop-dialog-container/desktop-dialog-container.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { FsDateModule } from '@firestitch/date';
import { FsDatePickerCalendarModule } from '../../../calendar/calendar.module';
import { FsDesktopCalendarDialogComponent } from './components/desktop-dialog/desktop-dialog.component';
import { FsDatePickerHeaderComponent } from './components/header/header.component';
import { FsDesktopDatePickerComponent } from './components/date-picker/date-picker.component';
import { FsMonthRangePickerComponent } from './components/month-range-picker/month-range-picker.component';
import { FsDatePickerHeaderMonthRangeComponent } from './components/header-month-range/header-month-range.component';
import * as i0 from "@angular/core";
export class FsDatePickerDesktopDialogContainerModule {
}
FsDatePickerDesktopDialogContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDesktopDialogContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDatePickerDesktopDialogContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDesktopDialogContainerModule, declarations: [FsDesktopCalendarDialogComponent,
        FsDatePickerHeaderComponent,
        FsDesktopDatePickerComponent,
        FsMonthRangePickerComponent,
        FsDatePickerHeaderMonthRangeComponent], imports: [CommonModule,
        MatButtonModule,
        MatIconModule,
        FsDatePickerCalendarModule,
        FsDateModule,
        ScrollingModule], exports: [FsDesktopCalendarDialogComponent] });
FsDatePickerDesktopDialogContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDesktopDialogContainerModule, imports: [[
            CommonModule,
            MatButtonModule,
            MatIconModule,
            FsDatePickerCalendarModule,
            FsDateModule,
            ScrollingModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDesktopDialogContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatButtonModule,
                        MatIconModule,
                        FsDatePickerCalendarModule,
                        FsDateModule,
                        ScrollingModule,
                    ],
                    declarations: [
                        FsDesktopCalendarDialogComponent,
                        FsDatePickerHeaderComponent,
                        FsDesktopDatePickerComponent,
                        FsMonthRangePickerComponent,
                        FsDatePickerHeaderMonthRangeComponent,
                    ],
                    exports: [
                        FsDesktopCalendarDialogComponent,
                    ],
                }]
        }] });
