// @firestitch/datepicker/esm2015/libs/dialog/modules/scroll-picker-dialog-container/scroll-picker-dialog-container.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { FsScrollPickerModule } from '@firestitch/scroll-picker';
import { FsDateScrollPickerMobileDialogComponent } from './components/date-scroll-picker-mobile-dialog/date-scroll-picker-mobile-dialog.component';
import { FsDateScrollPickerDesktopComponent } from './components/date-scroll-picker-desktop/date-scroll-picker-desktop.component';
import { FsDateScrollPickerDialogComponent } from './components/date-scroll-picker/date-scroll-picker.component';
import { MatListModule } from '@angular/material/list';
import * as i0 from "@angular/core";
export class FsDatePickerScrollPickerDialogContainerModule {
}
FsDatePickerScrollPickerDialogContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerScrollPickerDialogContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDatePickerScrollPickerDialogContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerScrollPickerDialogContainerModule, declarations: [FsDateScrollPickerDesktopComponent,
        FsDateScrollPickerMobileDialogComponent,
        FsDateScrollPickerDialogComponent], imports: [CommonModule,
        FormsModule,
        MatButtonModule,
        FsScrollPickerModule,
        MatListModule], exports: [FsDateScrollPickerDesktopComponent,
        FsDateScrollPickerMobileDialogComponent] });
FsDatePickerScrollPickerDialogContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerScrollPickerDialogContainerModule, imports: [[
            CommonModule,
            FormsModule,
            MatButtonModule,
            FsScrollPickerModule,
            MatListModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerScrollPickerDialogContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        MatButtonModule,
                        FsScrollPickerModule,
                        MatListModule,
                    ],
                    declarations: [
                        FsDateScrollPickerDesktopComponent,
                        FsDateScrollPickerMobileDialogComponent,
                        FsDateScrollPickerDialogComponent,
                    ],
                    exports: [
                        FsDateScrollPickerDesktopComponent,
                        FsDateScrollPickerMobileDialogComponent,
                    ],
                }]
        }] });
