// @firestitch/datepicker/esm2015/libs/dialog/modules/mobile-dialog-container/components/header-month-range/header-month-range.component.js
import { ChangeDetectionStrategy, Component, Input, ElementRef, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/icon";
import * as i2 from "@angular/common";
import * as i3 from "@firestitch/date";
export class FsDatePickerMobileHeaderMonthRangeComponent {
    constructor(_elRef) {
        this._elRef = _elRef;
    }
}
FsDatePickerMobileHeaderMonthRangeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileHeaderMonthRangeComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
FsDatePickerMobileHeaderMonthRangeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDatePickerMobileHeaderMonthRangeComponent, selector: "fs-date-picker-mobile-header-month-range", inputs: { rangeFrom: "rangeFrom", rangeTo: "rangeTo" }, ngImport: i0, template: "<ng-container *ngIf=\"rangeFrom || rangeTo else notSelected\">\n  <ng-container *ngIf=\"rangeFrom else notSelected\">\n    {{ rangeFrom | fsDateFormat: 'day-date' }}\n  </ng-container>\n  &nbsp;\n  <ng-container *ngIf=\"rangeTo\">\n    <mat-icon>arrow_forward</mat-icon>&nbsp;\n    {{ rangeTo | fsDateFormat: 'day-date' }}\n  </ng-container>\n</ng-container>\n\n\n<ng-template #notSelected>\n  Not Selected\n</ng-template>\n", styles: [":host{display:flex;justify-content:center;align-items:center}\n"], components: [{ type: i1.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "fsDateFormat": i3.FsDateFormatPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileHeaderMonthRangeComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-date-picker-mobile-header-month-range',
                    templateUrl: './header-month-range.component.html',
                    styleUrls: [
                        './header-month-range.component.scss',
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { rangeFrom: [{
                type: Input
            }], rangeTo: [{
                type: Input
            }] } });
