// @firestitch/date/esm2015/app/services/date-ago-factory.service.js
import { ComponentFactoryResolver, Injectable, Inject } from '@angular/core';
import { FsDateAgoComponent } from '../components/date-ago/date-ago.component';
import * as i0 from "@angular/core";
export class FsDateAgoFactory {
    constructor(factoryResolver) {
        this.factoryResolver = null;
        this.rootViewContainer = null;
        this.factoryResolver = factoryResolver;
    }
    setRootViewContainerRef(viewContainerRef) {
        this.rootViewContainer = viewContainerRef;
    }
    addDynamicComponent() {
        const factory = this.factoryResolver
            .resolveComponentFactory(FsDateAgoComponent);
        const component = factory
            .create(this.rootViewContainer.injector);
        this.rootViewContainer.insert(component.hostView);
        return component;
    }
}
FsDateAgoFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateAgoFactory, deps: [{ token: ComponentFactoryResolver }], target: i0.ɵɵFactoryTarget.Injectable });
FsDateAgoFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateAgoFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateAgoFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [ComponentFactoryResolver]
                }] }]; } });
