// @firestitch/dialog/esm2015/app/serivces/fs-dialog-router.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { MatDialog } from '@angular/material/dialog';
import { BehaviorSubject } from 'rxjs';
import { filter, map, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@angular/material/dialog";
export class FsDialogRouter {
    constructor(_router, _dialog) {
        this._router = _router;
        this._dialog = _dialog;
        this._activeDialogs = new WeakMap();
        this._dialogOpened$ = new BehaviorSubject(null);
    }
    openDialogForRoute(component, config) {
        const dialog = this.openDialog(component, config);
        this._registerDialog(component, dialog);
        return dialog;
    }
    navigate(commands, extras) {
        return this._router.navigate(commands, extras);
    }
    openDialog(component, config) {
        const dialog = this._dialog.open(component, config);
        this._registerDialog(component, dialog);
        return dialog;
    }
    dialogRef$(component) {
        return this._dialogOpened$
            .pipe(filter(() => {
            return this._activeDialogs.has(component);
        }), map(() => this._activeDialogs.get(component)));
    }
    _registerDialog(component, ref) {
        if (!this._activeDialogs.has(component)) {
            this._activeDialogs.set(component, ref);
            this._listenDialogClose(component, ref);
            this._dialogOpened$.next();
        }
    }
    _listenDialogClose(component, ref) {
        ref.afterClosed()
            .pipe(take(1))
            .subscribe(() => {
            this._activeDialogs.delete(component);
        });
    }
}
FsDialogRouter.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogRouter, deps: [{ token: i1.Router }, { token: i2.MatDialog }], target: i0.ɵɵFactoryTarget.Injectable });
FsDialogRouter.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogRouter, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogRouter, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.MatDialog }]; } });
