// @firestitch/dialog/esm2015/app/serivces/fs-dialog.js
import { Injectable } from '@angular/core';
import { FsDialogRouter } from './fs-dialog-router';
import * as i0 from "@angular/core";
import * as i1 from "./fs-dialog-router";
export class FsDialog {
    constructor(_dialogRouter) {
        this._dialogRouter = _dialogRouter;
    }
    open(component, config) {
        return this._dialogRouter.openDialog(component, config);
    }
    navigate(commands, extras) {
        return this._dialogRouter.navigate(commands, extras);
    }
    dialogRef$(component) {
        return this._dialogRouter.dialogRef$(component);
    }
}
FsDialog.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialog, deps: [{ token: i1.FsDialogRouter }], target: i0.ɵɵFactoryTarget.Injectable });
FsDialog.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialog, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialog, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.FsDialogRouter }]; } });
