// @firestitch/scroll-picker/esm2015/app/fs-scroll-picker.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScrollPickerComponent } from './components/scroll-picker/scroll-picker.component';
import { ScrollPickerTemplateComponent } from './directives/scroll-picker-template.directive';
import * as i0 from "@angular/core";
export class FsScrollPickerModule {
}
FsScrollPickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsScrollPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsScrollPickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsScrollPickerModule, declarations: [ScrollPickerComponent,
        ScrollPickerTemplateComponent], imports: [CommonModule], exports: [ScrollPickerComponent,
        ScrollPickerTemplateComponent] });
FsScrollPickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsScrollPickerModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsScrollPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        ScrollPickerComponent,
                        ScrollPickerTemplateComponent
                    ],
                    declarations: [
                        ScrollPickerComponent,
                        ScrollPickerTemplateComponent
                    ],
                }]
        }] });
