// @firestitch/datepicker/esm2015/app/components/range-picker/base/range-picker-from.component.js
import { ChangeDetectorRef, Directive, ElementRef, Injector, Optional, Self, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { startOfDay } from 'date-fns';
import { FsDatePickerDialogFactory } from '../../../../libs/dialog/services/dialog-factory.service';
import { PickerViewType } from '../../../../libs/common/enums/picker-view-type.enum';
import { RangePickerComponent } from '../base/range-picker-base.component';
import { FsRangePickerStoreService } from '../../../services/range-picker-store.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../../libs/dialog/services/dialog-factory.service";
import * as i2 from "@angular/forms";
import * as i3 from "../../../services/range-picker-store.service";
export class RangePickerFromComponent extends RangePickerComponent {
    constructor(_elRef, _injector, _datepickerFactory, _cdRef, _ngControl, _rangePickerStore) {
        super(_elRef, _injector, _datepickerFactory, 'from', _cdRef, _ngControl);
        this._elRef = _elRef;
        this._injector = _injector;
        this._datepickerFactory = _datepickerFactory;
        this._cdRef = _cdRef;
        this._ngControl = _ngControl;
        this._rangePickerStore = _rangePickerStore;
    }
    ngOnInit() {
        this.registerPicker();
        super.ngOnInit();
    }
    ngOnDestroy() {
        this._rangePickerStore.destroyStartDatePicker(this.name);
        this._destroy$.next();
        this._destroy$.complete();
    }
    registerPicker() {
        this._pickerRef = this._rangePickerStore.registerPickerFrom(this.name, this.value, this.view);
    }
    writeValue(value) {
        // Hot fix while angular has ongoing issue
        // https://github.com/angular/angular/issues/29218
        if (!this.onChange) {
            return;
        }
        super.writeValue(value);
        const [valuesAreDates] = this._checkValuesEquality(this.value, this._pickerRef.startDate);
        if ((valuesAreDates && !this._pickerRef.sameAsStartDate(this.value)) || (!valuesAreDates)) {
            this._pickerRef.updateStartDate(this.value);
        }
    }
    cleared(event) {
        event.stopPropagation();
        event.preventDefault();
        this.writeValue(null);
        this.onChange(this.value);
        this.onTouch(this.value);
    }
    /**
     * Set value which was selected in dialog
     * @param value
     */
    updateValueFromDialog(value) {
        this.updateValue(value);
        super.updateValueFromDialog(value);
    }
    updateValue(value) {
        if (this.view === PickerViewType.Date) {
            value = startOfDay(value);
        }
        this._pickerRef.updateStartDate(value);
        super.updateValue(value);
    }
    _tzChanged(originDate) {
        var _a;
        super._tzChanged(originDate);
        (_a = this._pickerRef) === null || _a === void 0 ? void 0 : _a.updateStartDate(this.value);
    }
    _processInputDate(date) {
        date = super._processInputDate(date);
        if (this.view === PickerViewType.Date && date) {
            date = startOfDay(date);
        }
        return date;
    }
}
RangePickerFromComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: RangePickerFromComponent, deps: [{ token: i0.ElementRef }, { token: i0.Injector }, { token: i1.FsDatePickerDialogFactory }, { token: i0.ChangeDetectorRef }, { token: i2.NgControl, optional: true, self: true }, { token: i3.FsRangePickerStoreService }], target: i0.ɵɵFactoryTarget.Directive });
RangePickerFromComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: RangePickerFromComponent, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: RangePickerFromComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Injector }, { type: i1.FsDatePickerDialogFactory }, { type: i0.ChangeDetectorRef }, { type: i2.NgControl, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: i3.FsRangePickerStoreService }]; } });
