// @firestitch/datepicker/esm2015/libs/calendar/calendar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatIconModule } from '@angular/material/icon';
import { FsDatePickerCalendarComponent } from './components/calendar/calendar.component';
import { FsDatePickerTimeComponent } from './components/time/time.component';
import * as i0 from "@angular/core";
export class FsDatePickerCalendarModule {
}
FsDatePickerCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDatePickerCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerCalendarModule, declarations: [FsDatePickerCalendarComponent,
        FsDatePickerTimeComponent], imports: [CommonModule,
        MatIconModule], exports: [FsDatePickerCalendarComponent,
        FsDatePickerTimeComponent] });
FsDatePickerCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerCalendarModule, imports: [[
            CommonModule,
            MatIconModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatIconModule,
                    ],
                    declarations: [
                        FsDatePickerCalendarComponent,
                        FsDatePickerTimeComponent,
                    ],
                    exports: [
                        FsDatePickerCalendarComponent,
                        FsDatePickerTimeComponent,
                    ],
                }]
        }] });
