// @firestitch/autocomplete-chips/esm2015/app/fs-autocomplete-chips.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatChipsModule } from '@angular/material/chips';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatRadioModule } from '@angular/material/radio';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { FsLabelModule } from '@firestitch/label';
import { FsAutocompleteChipsComponent } from './components/autocomplete-chips/autocomplete-chips.component';
import { FsAutocompleteObjectDirective } from './directives/autocomplete-object/autocomplete-object.directive';
import { FsAutocompleteChipsNoResultsDirective } from './directives/autocomplete-no-results/autocomplete-no-results.directive';
import { FsAutocompleteChipsStaticDirective } from './directives/static-template/static-template.directive';
import { FsAutocompleteChipSuffixDirective } from './directives/chip-suffix/chip-suffix.directive';
import { FsAutocompleteChipsSuffixDirective } from './directives/chips-suffix/chips-suffix.directive';
import * as i0 from "@angular/core";
export class FsAutocompleteChipsModule {
    static forRoot() {
        return {
            ngModule: FsAutocompleteChipsModule,
        };
    }
}
FsAutocompleteChipsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteChipsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsAutocompleteChipsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteChipsModule, declarations: [FsAutocompleteObjectDirective,
        FsAutocompleteChipsComponent,
        FsAutocompleteChipsStaticDirective,
        FsAutocompleteChipsNoResultsDirective,
        FsAutocompleteChipSuffixDirective,
        FsAutocompleteChipsSuffixDirective], imports: [CommonModule,
        FormsModule,
        DragDropModule,
        MatAutocompleteModule,
        MatChipsModule,
        MatInputModule,
        MatIconModule,
        MatRadioModule,
        FsLabelModule], exports: [FsAutocompleteChipsComponent,
        FsAutocompleteObjectDirective,
        FsAutocompleteChipsStaticDirective,
        FsAutocompleteChipsNoResultsDirective,
        FsAutocompleteChipSuffixDirective,
        FsAutocompleteChipsSuffixDirective] });
FsAutocompleteChipsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteChipsModule, imports: [[
            CommonModule,
            FormsModule,
            DragDropModule,
            MatAutocompleteModule,
            MatChipsModule,
            MatInputModule,
            MatIconModule,
            MatRadioModule,
            FsLabelModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteChipsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        DragDropModule,
                        MatAutocompleteModule,
                        MatChipsModule,
                        MatInputModule,
                        MatIconModule,
                        MatRadioModule,
                        FsLabelModule,
                    ],
                    exports: [
                        FsAutocompleteChipsComponent,
                        FsAutocompleteObjectDirective,
                        FsAutocompleteChipsStaticDirective,
                        FsAutocompleteChipsNoResultsDirective,
                        FsAutocompleteChipSuffixDirective,
                        FsAutocompleteChipsSuffixDirective,
                    ],
                    declarations: [
                        FsAutocompleteObjectDirective,
                        FsAutocompleteChipsComponent,
                        FsAutocompleteChipsStaticDirective,
                        FsAutocompleteChipsNoResultsDirective,
                        FsAutocompleteChipSuffixDirective,
                        FsAutocompleteChipsSuffixDirective,
                    ],
                }]
        }] });
