// @firestitch/datepicker/esm2015/libs/dialog/modules/mobile-dialog-container/components/time-picker/time-picker.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { MatBottomSheetRef } from '@angular/material/bottom-sheet';
import { FsDatePickerDialogModel } from '../../../../../dialog/classes/dialog-model';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/bottom-sheet";
import * as i2 from "../../../../../calendar/components/time/time.component";
import * as i3 from "@angular/common";
export class FsMobileTimePickerComponent {
    constructor(_bottomSheet) {
        this._bottomSheet = _bottomSheet;
        this.autoClose = true;
        this.showNotSelected = true;
    }
    dateChanged(date) {
        this.datePickerModel.model = date;
    }
    close() {
        if (this.autoClose) {
            this._bottomSheet.dismiss();
        }
    }
}
FsMobileTimePickerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMobileTimePickerComponent, deps: [{ token: i1.MatBottomSheetRef }], target: i0.ɵɵFactoryTarget.Component });
FsMobileTimePickerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsMobileTimePickerComponent, selector: "fs-datepicker-mobile-timepicker", inputs: { datePickerModel: "datePickerModel", autoClose: "autoClose", showNotSelected: "showNotSelected" }, ngImport: i0, template: "<fs-date-picker-time\n  class=\"fs-date-picker-dialog\"\n  [date]=\"datePickerModel.model$ | async\"\n  [minutes]=\"datePickerModel.minutes\"\n  [disabledTimes]=\"datePickerModel.disabledTimes$ | async\"\n  [expanded]=\"datePickerModel.timeExpanded$ | async\"\n  [showNotSelected]=\"showNotSelected\"\n  (onChange)=\"dateChanged($event)\"\n></fs-date-picker-time>\n", styles: [":host ::ng-deep .month-year.time{display:none}\n"], components: [{ type: i2.FsDatePickerTimeComponent, selector: "fs-date-picker-time", inputs: ["date", "disabledMinutes", "disabledHours", "disabledTimes", "expanded", "minutes", "showNotSelected"], outputs: ["onChange"] }], pipes: { "async": i3.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMobileTimePickerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-datepicker-mobile-timepicker',
                    templateUrl: './time-picker.component.html',
                    styleUrls: ['./time-picker.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.MatBottomSheetRef }]; }, propDecorators: { datePickerModel: [{
                type: Input
            }], autoClose: [{
                type: Input
            }], showNotSelected: [{
                type: Input
            }] } });
