// @firestitch/dialog/esm2015/app/components/fs-dialog/fs-dialog.component.js
import { ChangeDetectionStrategy, Component, Input, Optional, SkipSelf } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/dialog";
export class FsDialogComponent {
    constructor(_dialogRef) {
        this._dialogRef = _dialogRef;
        this.mobileMode = 'full';
    }
    ngAfterContentInit() {
        const mobileMode = 'fs-dialog-mobile-mode-' + this.mobileMode;
        if (this._dialogRef) {
            const pane = this._dialogRef._overlayRef.overlayElement;
            if (pane) {
                pane.classList.add('fs-dialog-overlay-pane');
                pane.classList.add(mobileMode);
            }
            const backdrop = this._dialogRef._overlayRef.backdropElement;
            if (backdrop) {
                backdrop.classList.add('fs-dialog-overlay-backdrop');
                backdrop.classList.add(mobileMode);
            }
        }
        window.document.body.classList.add('fs-dialog-open', mobileMode);
    }
    ngOnDestroy() {
        window.document.body.classList.remove('fs-dialog-open', `fs-dialog-mobile-mode-${this.mobileMode}`);
    }
}
FsDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogComponent, deps: [{ token: i1.MatDialogRef, optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.Component });
FsDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDialogComponent, selector: "fs-dialog", inputs: { mobileMode: "mobileMode" }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-dialog',
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.MatDialogRef, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }] }]; }, propDecorators: { mobileMode: [{
                type: Input,
                args: ['mobileMode']
            }] } });
