// @firestitch/autocomplete/esm2015/app/directives/autocomplete-static/autocomplete-static.directive.js
import { Directive, Output, EventEmitter, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class FsAutocompleteStaticDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.showWhenKeyword = false;
        this.selected = new EventEmitter();
    }
}
FsAutocompleteStaticDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteStaticDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
FsAutocompleteStaticDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsAutocompleteStaticDirective, selector: "[fsAutocompleteStatic],[fsAutocompleteStaticTemplate]", inputs: { showWhenKeyword: "showWhenKeyword" }, outputs: { selected: "selected" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteStaticDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsAutocompleteStatic],[fsAutocompleteStaticTemplate]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; }, propDecorators: { showWhenKeyword: [{
                type: Input
            }], selected: [{
                type: Output
            }] } });
