// @firestitch/datepicker/esm2015/app/components/date-range-separator/date-range-separator.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class DateRangeSeparatorComponent {
}
DateRangeSeparatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangeSeparatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DateRangeSeparatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: DateRangeSeparatorComponent, selector: "fs-date-range-separator", ngImport: i0, template: 'to', isInline: true, styles: [":host{display:inline-block;padding:0 10px}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangeSeparatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-date-range-separator',
                    template: 'to',
                    styleUrls: ['./date-range-separator.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
