// @firestitch/date/esm2015/app/fs-date.module.js
import { NgModule } from '@angular/core';
import { MatTooltipModule } from '@angular/material/tooltip';
import { CommonModule } from '@angular/common';
import { FsCommonModule } from '@firestitch/common';
import { FsDatePipe } from './pipes/date.pipe';
import { FsDateRangePipe } from './pipes/date-range.pipe';
import { FsDateFormatPipe } from './pipes/date-format.pipe';
import { FsDateDurationPipe } from './pipes/date-duration.pipe';
import { FsDateAgoPipe } from './pipes/date-ago.pipe';
import { FsDateAgoDirective } from './directives/fs-date-ago.directive';
import { FsDateAgoComponent } from './components/date-ago/date-ago.component';
import { FsDateAgoFactory } from './services/date-ago-factory.service';
import * as i0 from "@angular/core";
export class FsDateModule {
    static forRoot() {
        return {
            ngModule: FsDateModule,
            providers: [
                FsDateAgoFactory
            ]
        };
    }
}
FsDateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateModule, declarations: [FsDatePipe,
        FsDateAgoPipe,
        FsDateDurationPipe,
        FsDateFormatPipe,
        FsDateRangePipe,
        FsDateAgoDirective,
        FsDateAgoComponent], imports: [CommonModule,
        MatTooltipModule,
        FsCommonModule], exports: [FsDatePipe,
        FsDateAgoPipe,
        FsDateDurationPipe,
        FsDateFormatPipe,
        FsDateRangePipe,
        FsDateAgoDirective,
        FsDateAgoComponent] });
FsDateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateModule, providers: [], imports: [[
            CommonModule,
            MatTooltipModule,
            FsCommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatTooltipModule,
                        FsCommonModule
                    ],
                    exports: [
                        FsDatePipe,
                        FsDateAgoPipe,
                        FsDateDurationPipe,
                        FsDateFormatPipe,
                        FsDateRangePipe,
                        FsDateAgoDirective,
                        FsDateAgoComponent
                    ],
                    entryComponents: [
                        FsDateAgoComponent
                    ],
                    declarations: [
                        FsDatePipe,
                        FsDateAgoPipe,
                        FsDateDurationPipe,
                        FsDateFormatPipe,
                        FsDateRangePipe,
                        FsDateAgoDirective,
                        FsDateAgoComponent
                    ],
                    providers: [],
                }]
        }] });
