// @firestitch/datepicker/esm2015/app/components/date-picker-trigger/date-picker-trigger.component.js
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Optional, Output } from '@angular/core';
import { MatFormField } from '@angular/material/form-field';
import { PickerViewType } from '../../../libs/common/enums/picker-view-type.enum';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/form-field";
import * as i2 from "@angular/material/icon";
import * as i3 from "@angular/common";
export class FsDatePickerTriggerComponent {
    constructor(matFormField, el) {
        this.matFormField = matFormField;
        this.el = el;
        this.click = new EventEmitter();
    }
    get isTimeView() {
        return this.view === PickerViewType.Time;
    }
    ngAfterViewInit() {
        var _a;
        const matElementRef = this.matFormField._elementRef.nativeElement;
        const infixEl = matElementRef.querySelector('.mat-form-field-infix');
        (_a = matElementRef
            .querySelector('.mat-form-field-flex')) === null || _a === void 0 ? void 0 : _a.insertBefore(this.el.nativeElement.firstChild, infixEl);
    }
    triggerClick(event) {
        event.preventDefault();
        event.stopPropagation();
        this.click.emit();
    }
}
FsDatePickerTriggerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerTriggerComponent, deps: [{ token: i1.MatFormField, optional: true }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
FsDatePickerTriggerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: { disabled: "disabled", view: "view" }, outputs: { click: "click" }, ngImport: i0, template: "<div class=\"mat-form-field-preffix fs-date-picker-trigger\" [ngClass]=\"{ disabled: disabled }\">\n  <mat-icon (click)=\"triggerClick($event)\">\n    <ng-container *ngIf=\"isTimeView; else calendarIcon\">\n      schedule\n    </ng-container>\n    <ng-template #calendarIcon>\n      calendar_today\n    </ng-template>\n  </mat-icon>\n</div>\n", styles: [".fs-date-picker-trigger{margin-right:5px;margin-bottom:2px;align-self:flex-end;-webkit-user-select:none;user-select:none}.fs-date-picker-trigger mat-icon{cursor:pointer}.disabled{color:#00000061;pointer-events:none}\n"], components: [{ type: i2.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }], directives: [{ type: i3.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerTriggerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-datepicker-trigger',
                    templateUrl: 'date-picker-trigger.component.html',
                    styleUrls: ['date-picker-trigger.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.MatFormField, decorators: [{
                    type: Optional
                }] }, { type: i0.ElementRef }]; }, propDecorators: { disabled: [{
                type: Input
            }], view: [{
                type: Input
            }], click: [{
                type: Output
            }] } });
