// @firestitch/date/esm2015/app/directives/fs-date-ago.directive.js
import { Directive, ElementRef, Inject, Input, Renderer2, ViewContainerRef } from '@angular/core';
import { FsDateAgoFactory } from '../services/date-ago-factory.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/date-ago-factory.service";
export class FsDateAgoDirective {
    constructor(fsDateAgoFactory, renderer, elementRef, viewContainerRef) {
        this.fsDateAgoFactory = fsDateAgoFactory;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.viewContainerRef = viewContainerRef;
        this.date = null;
        this.showTime = false;
        this.format = 'date';
        this.dateAgoRef = null;
    }
    ngOnInit() {
    }
    ngOnChanges() {
        this.renderDateAgo();
    }
    renderDateAgo() {
        this.removeDateAgo();
        this.fsDateAgoFactory.setRootViewContainerRef(this.viewContainerRef);
        this.dateAgoRef = this.fsDateAgoFactory.addDynamicComponent();
        Object.assign(this.dateAgoRef.instance, {
            date: this.date,
            showTime: this.showTime,
            format: this.format
        });
    }
    removeDateAgo() {
        if (this.dateAgoRef) {
            this.renderer.removeChild(this.elementRef.nativeElement, this.dateAgoRef.instance.elementRef.nativeElement);
        }
    }
}
FsDateAgoDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateAgoDirective, deps: [{ token: i1.FsDateAgoFactory }, { token: i0.Renderer2 }, { token: i0.ElementRef }, { token: ViewContainerRef }], target: i0.ɵɵFactoryTarget.Directive });
FsDateAgoDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsDateAgoDirective, selector: "[fsDateAgo]", inputs: { date: "date", showTime: "showTime", format: "format" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateAgoDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsDateAgo]'
                }]
        }], ctorParameters: function () { return [{ type: i1.FsDateAgoFactory }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [ViewContainerRef]
                }] }]; }, propDecorators: { date: [{
                type: Input
            }], showTime: [{
                type: Input
            }], format: [{
                type: Input
            }] } });
