// @firestitch/datepicker/esm2015/libs/dialog/classes/dialog-ref.js
import { Subject } from 'rxjs';
import { skip } from 'rxjs/operators';
import { isEqual, forEach } from 'lodash-es';
import { FsDatePickerDialogModel } from './dialog-model';
import { FsDatePickerOverlayRef } from './overlay-ref';
export class FsDatePickerDialogRef {
    constructor(_pickerOptions) {
        this._pickerOptions = _pickerOptions;
        this._overlayRef = new FsDatePickerOverlayRef();
        this._close$ = new Subject();
        /**
         * Visual components. Can be changed by summary widget but only if _view allowed to do this.
         */
        this._componentsDefault = {
            calendarStart: false,
            calendarEnd: false,
            timeStart: false,
            timeEnd: false
        };
        this._components = null;
        this._init();
    }
    get pickerModel() {
        return this._dialogModel;
    }
    get options() {
        return this._pickerOptions;
    }
    get value$() {
        return this._value$;
    }
    get pickerOverlayRef() {
        return this._overlayRef;
    }
    get close$() {
        return this._close$.asObservable();
    }
    get components() {
        return this._components;
    }
    updateValue(value) {
        this._dialogModel.model = value;
    }
    close() {
        this.pickerOverlayRef.close();
        this._close$.next();
        this._close$.complete();
    }
    _init() {
        this._initComponents();
        this._initModel();
    }
    _initModel() {
        this._dialogModel = new FsDatePickerDialogModel(this._pickerOptions);
        this._initValue();
    }
    _initValue() {
        if (this.options.view === 'week') {
            this._value$ = this._dialogModel.period$
                .pipe(skip(1));
        }
        else {
            this._value$ = this._dialogModel.model$
                .pipe(skip(1));
        }
    }
    _initComponents() {
        const value = Object.assign(Object.assign({}, this._componentsDefault), this._pickerOptions.components || {});
        const tempData = Object.assign({}, value);
        const allowable = [];
        if (['week', 'date', 'datetime', 'monthrange'].indexOf(this._pickerOptions.view) !== -1) {
            allowable.push('calendarStart');
            allowable.push('calendarEnd');
        }
        if (['time', 'datetime'].indexOf(this._pickerOptions.view) !== -1) {
            allowable.push('timeStart');
            allowable.push('timeEnd');
        }
        forEach(tempData, (item, index) => {
            tempData[index] = allowable.indexOf(index) !== -1 ? item : false;
        });
        // Updating components only if all value object is valid
        if (isEqual(value, tempData)) {
            this._components = value;
        }
    }
}
