// @firestitch/dialog/bundles/firestitch-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/router'), require('@angular/material/dialog'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@firestitch/dialog', ['exports', '@angular/core', '@angular/common', '@angular/router', '@angular/material/dialog', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.firestitch = global.firestitch || {}, global.firestitch.dialog = {}), global.ng.core, global.ng.common, global.ng.router, global.ng.material.dialog, global.rxjs, global.rxjs.operators));
})(this, (function (exports, i0, common, i1$1, i1, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var FsDialogComponent = /** @class */ (function () {
        function FsDialogComponent(_dialogRef) {
            this._dialogRef = _dialogRef;
            this.mobileMode = 'full';
        }
        FsDialogComponent.prototype.ngAfterContentInit = function () {
            var mobileMode = 'fs-dialog-mobile-mode-' + this.mobileMode;
            if (this._dialogRef) {
                var pane = this._dialogRef._overlayRef.overlayElement;
                if (pane) {
                    pane.classList.add('fs-dialog-overlay-pane');
                    pane.classList.add(mobileMode);
                }
                var backdrop = this._dialogRef._overlayRef.backdropElement;
                if (backdrop) {
                    backdrop.classList.add('fs-dialog-overlay-backdrop');
                    backdrop.classList.add(mobileMode);
                }
            }
            window.document.body.classList.add('fs-dialog-open', mobileMode);
        };
        FsDialogComponent.prototype.ngOnDestroy = function () {
            window.document.body.classList.remove('fs-dialog-open', "fs-dialog-mobile-mode-" + this.mobileMode);
        };
        return FsDialogComponent;
    }());
    FsDialogComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogComponent, deps: [{ token: i1__namespace.MatDialogRef, optional: true, skipSelf: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    FsDialogComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDialogComponent, selector: "fs-dialog", inputs: { mobileMode: "mobileMode" }, ngImport: i0__namespace, template: '<ng-content></ng-content>', isInline: true, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'fs-dialog',
                        template: '<ng-content></ng-content>',
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.MatDialogRef, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.SkipSelf
                        }] }];
        }, propDecorators: { mobileMode: [{
                    type: i0.Input,
                    args: ['mobileMode']
                }] } });

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var FsDialogRouter = /** @class */ (function () {
        function FsDialogRouter(_router, _dialog) {
            this._router = _router;
            this._dialog = _dialog;
            this._activeDialogs = new WeakMap();
            this._dialogOpened$ = new rxjs.BehaviorSubject(null);
        }
        FsDialogRouter.prototype.openDialogForRoute = function (component, config) {
            var dialog = this.openDialog(component, config);
            this._registerDialog(component, dialog);
            return dialog;
        };
        FsDialogRouter.prototype.navigate = function (commands, extras) {
            return this._router.navigate(commands, extras);
        };
        FsDialogRouter.prototype.openDialog = function (component, config) {
            var dialog = this._dialog.open(component, config);
            this._registerDialog(component, dialog);
            return dialog;
        };
        FsDialogRouter.prototype.dialogRef$ = function (component) {
            var _this = this;
            return this._dialogOpened$
                .pipe(operators.filter(function () {
                return _this._activeDialogs.has(component);
            }), operators.map(function () { return _this._activeDialogs.get(component); }));
        };
        FsDialogRouter.prototype._registerDialog = function (component, ref) {
            if (!this._activeDialogs.has(component)) {
                this._activeDialogs.set(component, ref);
                this._listenDialogClose(component, ref);
                this._dialogOpened$.next();
            }
        };
        FsDialogRouter.prototype._listenDialogClose = function (component, ref) {
            var _this = this;
            ref.afterClosed()
                .pipe(operators.take(1))
                .subscribe(function () {
                _this._activeDialogs.delete(component);
            });
        };
        return FsDialogRouter;
    }());
    FsDialogRouter.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogRouter, deps: [{ token: i1__namespace$1.Router }, { token: i1__namespace.MatDialog }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    FsDialogRouter.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogRouter, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogRouter, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }], ctorParameters: function () { return [{ type: i1__namespace$1.Router }, { type: i1__namespace.MatDialog }]; } });

    var FsDialogRouteComponent = /** @class */ (function () {
        function FsDialogRouteComponent(_route, _router, _dialogRouter, _componentFactory, _viewContainerRef) {
            this._route = _route;
            this._router = _router;
            this._dialogRouter = _dialogRouter;
            this._componentFactory = _componentFactory;
            this._viewContainerRef = _viewContainerRef;
            this._hasActiveNavigation = false;
            this._destroy$ = new rxjs.Subject();
        }
        FsDialogRouteComponent.prototype.ngOnInit = function () {
            this._listenNavigationEvents();
            this.openDialog();
        };
        FsDialogRouteComponent.prototype.ngOnDestroy = function () {
            this._destroy$.next();
            this._destroy$.complete();
        };
        FsDialogRouteComponent.prototype.openDialog = function () {
            var _a;
            return __awaiter(this, void 0, void 0, function () {
                var dialogConfig, _c;
                return __generator(this, function (_d) {
                    switch (_d.label) {
                        case 0:
                            dialogConfig = Object.assign({}, (_a = this._route.snapshot.data) === null || _a === void 0 ? void 0 : _a.fsDialog);
                            if (!(dialogConfig === null || dialogConfig === void 0 ? void 0 : dialogConfig.component)) return [3 /*break*/, 4];
                            if (!(dialogConfig.component instanceof Promise)) return [3 /*break*/, 2];
                            _c = this;
                            return [4 /*yield*/, this._openLazyDialog(dialogConfig)];
                        case 1:
                            _c._dialog = _d.sent();
                            return [3 /*break*/, 3];
                        case 2:
                            this._dialog = this._openDialog(dialogConfig);
                            _d.label = 3;
                        case 3:
                            this._listenDialogClose();
                            _d.label = 4;
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        FsDialogRouteComponent.prototype._listenNavigationEvents = function () {
            var _this = this;
            this._router.events
                .pipe(operators.filter(function (event) {
                return event instanceof i1$1.NavigationStart;
            }), operators.takeUntil(this._destroy$))
                .subscribe(function (event) {
                _this._hasActiveNavigation = true;
            });
            this._router.events
                .pipe(operators.filter(function (event) {
                return event instanceof i1$1.NavigationEnd;
            }), operators.takeUntil(this._destroy$))
                .subscribe(function (event) {
                _this._hasActiveNavigation = false;
            });
        };
        FsDialogRouteComponent.prototype._listenDialogClose = function () {
            var _this = this;
            rxjs.merge(this._dialog.afterClosed())
                .pipe(operators.filter(function () { return !_this._hasActiveNavigation; }), operators.takeUntil(this._destroy$))
                .subscribe(function () {
                _this._navigateOutFromDialog();
            });
        };
        FsDialogRouteComponent.prototype._navigateOutFromDialog = function () {
            var _a, _b;
            var stepsBack = 1;
            var route = this._route.parent;
            // we are looking for parent route to navigate from current dialog
            // but we have to skip routes like loadChildren or redirectTo
            // because it does not make sence to do navigation to them
            while (route.routeConfig && (route.routeConfig.redirectTo || route.routeConfig.path === '') && route.parent) {
                stepsBack++;
                if (route.routeConfig.path === '' && ((_a = route.parent) === null || _a === void 0 ? void 0 : _a.routeConfig.loadChildren) && ((_b = route.parent) === null || _b === void 0 ? void 0 : _b.parent)) {
                    route = route.parent.parent;
                }
                else {
                    route = route.parent;
                }
            }
            // make relative navigation path like '../../../';
            var navigationPath = Array
                .from(Array(stepsBack), function () { return '../'; })
                .join('');
            // Do it!
            this._router.navigate([navigationPath], { relativeTo: this._route, queryParamsHandling: 'merge' });
        };
        FsDialogRouteComponent.prototype._openLazyDialog = function (dialogConfig) {
            return __awaiter(this, void 0, void 0, function () {
                var loadedComponent, componentName, factory;
                return __generator(this, function (_c) {
                    switch (_c.label) {
                        case 0: return [4 /*yield*/, dialogConfig.component];
                        case 1:
                            loadedComponent = _c.sent();
                            componentName = Object.keys(loadedComponent)[0];
                            if (!componentName) {
                                throw Error('Lazy loading dialog component error! Component not found!');
                            }
                            factory = this._componentFactory.resolveComponentFactory(loadedComponent[componentName]);
                            dialogConfig.component = factory.componentType;
                            return [2 /*return*/, this._openDialog(dialogConfig)];
                    }
                });
            });
        };
        FsDialogRouteComponent.prototype._openDialog = function (routeDialogConfig) {
            var dialogComponent = routeDialogConfig.component, dialogConfig = __rest(routeDialogConfig, ["component"]);
            dialogConfig.viewContainerRef = this._viewContainerRef;
            return this._dialogRouter.openDialogForRoute(dialogComponent, dialogConfig);
        };
        return FsDialogRouteComponent;
    }());
    FsDialogRouteComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogRouteComponent, deps: [{ token: i1__namespace$1.ActivatedRoute }, { token: i1__namespace$1.Router }, { token: FsDialogRouter }, { token: i0__namespace.ComponentFactoryResolver }, { token: i0__namespace.ViewContainerRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    FsDialogRouteComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDialogRouteComponent, selector: "ng-component", ngImport: i0__namespace, template: '<router-outlet></router-outlet>', isInline: true, directives: [{ type: i1__namespace$1.RouterOutlet, selector: "router-outlet", outputs: ["activate", "deactivate"], exportAs: ["outlet"] }] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogRouteComponent, decorators: [{
                type: i0.Component,
                args: [{
                        template: '<router-outlet></router-outlet>',
                    }]
            }], ctorParameters: function () { return [{ type: i1__namespace$1.ActivatedRoute }, { type: i1__namespace$1.Router }, { type: FsDialogRouter }, { type: i0__namespace.ComponentFactoryResolver }, { type: i0__namespace.ViewContainerRef }]; } });

    var FsDialogModule = /** @class */ (function () {
        function FsDialogModule() {
        }
        FsDialogModule.forRoot = function () {
            return {
                ngModule: FsDialogModule,
                providers: [
                    {
                        provide: i1.MAT_DIALOG_DEFAULT_OPTIONS,
                        useValue: {
                            panelClass: 'fs-dialog-overlay-pane',
                            maxWidth: '95vw',
                            disableClose: false,
                            autoFocus: true,
                            hasBackdrop: true
                        }
                    },
                ]
            };
        };
        return FsDialogModule;
    }());
    FsDialogModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    FsDialogModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogModule, declarations: [FsDialogComponent,
            FsDialogRouteComponent], imports: [common.CommonModule,
            i1$1.RouterModule], exports: [i1.MatDialogModule,
            FsDialogComponent,
            FsDialogRouteComponent] });
    FsDialogModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogModule, imports: [[
                common.CommonModule,
                i1$1.RouterModule,
            ], i1.MatDialogModule] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialogModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            common.CommonModule,
                            i1$1.RouterModule,
                        ],
                        declarations: [
                            FsDialogComponent,
                            FsDialogRouteComponent,
                        ],
                        exports: [
                            i1.MatDialogModule,
                            FsDialogComponent,
                            FsDialogRouteComponent,
                        ]
                    }]
            }] });

    var FsDialog = /** @class */ (function () {
        function FsDialog(_dialogRouter) {
            this._dialogRouter = _dialogRouter;
        }
        FsDialog.prototype.open = function (component, config) {
            return this._dialogRouter.openDialog(component, config);
        };
        FsDialog.prototype.navigate = function (commands, extras) {
            return this._dialogRouter.navigate(commands, extras);
        };
        FsDialog.prototype.dialogRef$ = function (component) {
            return this._dialogRouter.dialogRef$(component);
        };
        return FsDialog;
    }());
    FsDialog.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialog, deps: [{ token: FsDialogRouter }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    FsDialog.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialog, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsDialog, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }], ctorParameters: function () { return [{ type: FsDialogRouter }]; } });

    /*
     * Public API Surface of fs-menu
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.FsDialog = FsDialog;
    exports.FsDialogComponent = FsDialogComponent;
    exports.FsDialogModule = FsDialogModule;
    exports.FsDialogRouteComponent = FsDialogRouteComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
