// @firestitch/datepicker/esm2015/app/classes/range-picker-ref.js
import { BehaviorSubject } from 'rxjs';
import { endOfDay, startOfDay } from 'date-fns';
import { PickerViewType } from '../../libs/common/enums/picker-view-type.enum';
import { isDateAfter } from '../../libs/common/helpers/is-date-after';
export class RangePickerRef {
    constructor(view) {
        this.view = view;
        this._startDatePickerExists = false;
        this._endDatePickerExists = false;
        this._startDate$ = new BehaviorSubject(null);
        this._endDate$ = new BehaviorSubject(null);
        this._startDate = null;
        this._endDate = null;
    }
    get startDate() {
        return this._startDate;
    }
    get endDate() {
        return this._endDate;
    }
    get startDate$() {
        return this._startDate$.asObservable();
    }
    get endDate$() {
        return this._endDate$.asObservable();
    }
    get startDatePickerExists() {
        return this._startDatePickerExists;
    }
    get endDatePickerExists() {
        return this._endDatePickerExists;
    }
    get isRangeValid() {
        return isDateAfter(this._endDate, this._startDate, this.view);
    }
    /**
     * Update start date and change end date if needed
     * @param value
     */
    updateStartDate(value) {
        this._startDatePickerExists = true;
        if (!!value
            && (this.view === PickerViewType.Date || this.view === PickerViewType.MonthRange)) {
            value = startOfDay(value);
        }
        this._startDate = value && new Date(value) || null;
        this._startDate$.next(this._startDate);
    }
    /**
     * Update end date
     * @param value
     */
    updateEndDate(value) {
        if (!!value
            && (this.view === PickerViewType.Date || this.view === PickerViewType.MonthRange)) {
            value = endOfDay(value);
        }
        this._endDate = value && new Date(value) || null;
        this._endDatePickerExists = true;
        this._endDate$.next(this._endDate);
    }
    sameAsStartDate(value) {
        if (!!value
            && (this.view === PickerViewType.Date || this.view === PickerViewType.MonthRange)) {
            value = startOfDay(value);
        }
        return this._startDate === value;
    }
    sameAsEndDate(value) {
        if (!!value
            && (this.view === PickerViewType.Date || this.view === PickerViewType.MonthRange)) {
            value = endOfDay(value);
        }
        return this._endDate === value;
    }
    /**
     * Mark start date picker as destroyed
     */
    destroyStartDatePicker() {
        this._startDatePickerExists = false;
    }
    /**
     * Mark end date picker as destroyed
     */
    destroyEndDatePicker() {
        this._endDatePickerExists = false;
    }
    /**
     * destroy everything related with picker
     */
    destroy() { }
}
