// @firestitch/dialog/esm2015/app/fs-dialog.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { MAT_DIALOG_DEFAULT_OPTIONS, MatDialogModule } from '@angular/material/dialog';
import { FsDialogComponent } from './components/fs-dialog/fs-dialog.component';
import { FsDialogRouteComponent } from './components/route/route.component';
import * as i0 from "@angular/core";
export class FsDialogModule {
    static forRoot() {
        return {
            ngModule: FsDialogModule,
            providers: [
                {
                    provide: MAT_DIALOG_DEFAULT_OPTIONS,
                    useValue: {
                        panelClass: 'fs-dialog-overlay-pane',
                        maxWidth: '95vw',
                        disableClose: false,
                        autoFocus: true,
                        hasBackdrop: true
                    }
                },
            ]
        };
    }
}
FsDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogModule, declarations: [FsDialogComponent,
        FsDialogRouteComponent], imports: [CommonModule,
        RouterModule], exports: [MatDialogModule,
        FsDialogComponent,
        FsDialogRouteComponent] });
FsDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogModule, imports: [[
            CommonModule,
            RouterModule,
        ], MatDialogModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                    ],
                    declarations: [
                        FsDialogComponent,
                        FsDialogRouteComponent,
                    ],
                    exports: [
                        MatDialogModule,
                        FsDialogComponent,
                        FsDialogRouteComponent,
                    ]
                }]
        }] });
