// @firestitch/datepicker/esm2015/app/components/range-picker/to/date-range-picker-to.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { FsDatePickerComponent } from '../../date-picker/date-picker.component';
import { RangePickerToComponent } from '../base/range-picker-to.component';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/clear";
import * as i2 from "../../date-picker-trigger/date-picker-trigger.component";
export class DateRangePickerToComponent extends RangePickerToComponent {
    set fsDateRangeTo(value) {
        this._name = value;
    }
    set fsDateRangeToPicker(value) {
        this._name = value;
    }
}
DateRangePickerToComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangePickerToComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
DateRangePickerToComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: DateRangePickerToComponent, selector: "[fsDateRangeTo],[fsDateRangeToPicker]", inputs: { fsDateRangeTo: "fsDateRangeTo", fsDateRangeToPicker: "fsDateRangeToPicker" }, usesInheritance: true, ngImport: i0, template: "\n    <fs-clear [show]=\"value && !disabled && !readonly && clear\" (clear)=\"cleared($event)\"></fs-clear>\n    <fs-datepicker-trigger (click)=\"triggerClick()\" [disabled]=\"disabled || readonly\" [view]=\"view\"></fs-datepicker-trigger>\n  ", isInline: true, components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }, { type: i2.FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: ["disabled", "view"], outputs: ["click"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangePickerToComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsDateRangeTo],[fsDateRangeToPicker]',
                    template: FsDatePickerComponent.template,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { fsDateRangeTo: [{
                type: Input
            }], fsDateRangeToPicker: [{
                type: Input
            }] } });
