// @firestitch/datepicker/esm2015/app/components/range-picker/to/month-range-picker-to.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { takeUntil } from 'rxjs/operators';
import { zonedTimeToUtc } from 'date-fns-tz';
import { endOfDay } from 'date-fns';
import { PickerViewType } from '../../../../libs/common/enums/picker-view-type.enum';
import { FsDatePickerComponent } from '../../date-picker/date-picker.component';
import { RangePickerToComponent } from '../base/range-picker-to.component';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/clear";
import * as i2 from "../../date-picker-trigger/date-picker-trigger.component";
export class MonthRangePickerToComponent extends RangePickerToComponent {
    constructor() {
        super(...arguments);
        this.view = PickerViewType.MonthRange;
        this._skipUpdateValue = false;
    }
    set fsMonthRangeTo(value) {
        this._name = value;
    }
    set fsMonthRangeToPicker(value) {
        this._name = value;
    }
    updateValue(value) {
        if (value) {
            value = endOfDay(value);
        }
        this._value = value;
        this.updateInput(this._value);
        if (value && this.timezone) {
            value = zonedTimeToUtc(value, this.timezone);
        }
        this.onChange(value);
        this.onTouch(value);
    }
    _tzChanged(originDate) {
        this._skipUpdateValue = true;
        super._tzChanged(originDate);
    }
    _processInputDate(date) {
        date = super._processInputDate(date);
        if (date) {
            date = endOfDay(date);
        }
        return date;
    }
    /**
     * When it's month range picker - we have to listen changes over rangePickerRef
     */
    _listenDialogValueChanges() {
        // NOOP
    }
    _subscribeToPickerRefUpdates() {
        this._pickerRefUpdates$(this._pickerRef.endDate$)
            .pipe(takeUntil(this._destroy$))
            .subscribe((newValue) => {
            if (this._skipUpdateValue) {
                this._skipUpdateValue = false;
                return;
            }
            this.updateValue(newValue);
            this._ngControl.control.markAsDirty();
            this._ngControl.control.updateValueAndValidity();
            this._cdRef.markForCheck();
        });
    }
}
MonthRangePickerToComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MonthRangePickerToComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
MonthRangePickerToComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: MonthRangePickerToComponent, selector: "[fsMonthRangeTo],[fsMonthRangeToPicker]", inputs: { fsMonthRangeTo: "fsMonthRangeTo", fsMonthRangeToPicker: "fsMonthRangeToPicker" }, usesInheritance: true, ngImport: i0, template: "\n    <fs-clear [show]=\"value && !disabled && !readonly && clear\" (clear)=\"cleared($event)\"></fs-clear>\n    <fs-datepicker-trigger (click)=\"triggerClick()\" [disabled]=\"disabled || readonly\" [view]=\"view\"></fs-datepicker-trigger>\n  ", isInline: true, components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }, { type: i2.FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: ["disabled", "view"], outputs: ["click"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MonthRangePickerToComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsMonthRangeTo],[fsMonthRangeToPicker]',
                    template: FsDatePickerComponent.template,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { fsMonthRangeTo: [{
                type: Input
            }], fsMonthRangeToPicker: [{
                type: Input
            }] } });
