// @firestitch/datepicker/esm2015/app/components/range-picker/to/time-range-picker-to.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { PickerViewType } from '../../../../libs/common/enums/picker-view-type.enum';
import { FsDatePickerComponent } from '../../date-picker/date-picker.component';
import { RangePickerToComponent } from '../base/range-picker-to.component';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/clear";
import * as i2 from "../../date-picker-trigger/date-picker-trigger.component";
export class TimeRangePickerToComponent extends RangePickerToComponent {
    constructor() {
        super(...arguments);
        this.view = PickerViewType.Time;
    }
    set fsTimeRangeTo(value) {
        this._name = value;
    }
    set fsTimeRangeToPicker(value) {
        this._name = value;
    }
}
TimeRangePickerToComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TimeRangePickerToComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TimeRangePickerToComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: TimeRangePickerToComponent, selector: "[fsTimeRangeTo],[fsTimeRangeToPicker]", inputs: { fsTimeRangeTo: "fsTimeRangeTo", fsTimeRangeToPicker: "fsTimeRangeToPicker", view: "view" }, usesInheritance: true, ngImport: i0, template: "\n    <fs-clear [show]=\"value && !disabled && !readonly && clear\" (clear)=\"cleared($event)\"></fs-clear>\n    <fs-datepicker-trigger (click)=\"triggerClick()\" [disabled]=\"disabled || readonly\" [view]=\"view\"></fs-datepicker-trigger>\n  ", isInline: true, components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }, { type: i2.FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: ["disabled", "view"], outputs: ["click"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TimeRangePickerToComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsTimeRangeTo],[fsTimeRangeToPicker]',
                    template: FsDatePickerComponent.template,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { fsTimeRangeTo: [{
                type: Input
            }], fsTimeRangeToPicker: [{
                type: Input
            }], view: [{
                type: Input
            }] } });
