// @firestitch/form/esm2015/app/directives/validators/compare.directive.js
import { Directive, Input } from '@angular/core';
import { FsControlDirective } from './control.directive';
import { VALIDATE_MESSAGE_PROVIDER } from '../../providers/validate-messages.provider';
import * as i0 from "@angular/core";
export class FsFormCompareDirective extends FsControlDirective {
    set validationMessage(value) {
        this._validateMessages.compare = value;
    }
    ngOnChanges() {
        this._control.updateValueAndValidity();
    }
    validate(control) {
        if (this.fsFormCompare.value === this.elementRef.nativeElement.value) {
            return null;
        }
        else {
            return { compare: true };
        }
    }
    ngAfterViewInit() {
        this.fsFormCompare.addEventListener('keyup', () => {
            this._control.updateValueAndValidity();
        }, false);
    }
    ngOnDestroy() {
        this.fsFormCompare.removeEventListener('keyup', () => {
            this._control.updateValueAndValidity();
        }, false);
    }
}
FsFormCompareDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormCompareDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
FsFormCompareDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsFormCompareDirective, selector: "[fsFormCompare]", inputs: { fsFormCompare: "fsFormCompare", validationMessage: ["fsFormCompareMessage", "validationMessage"] }, providers: [
        VALIDATE_MESSAGE_PROVIDER,
    ], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormCompareDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsFormCompare]',
                    providers: [
                        VALIDATE_MESSAGE_PROVIDER,
                    ],
                }]
        }], propDecorators: { fsFormCompare: [{
                type: Input
            }], validationMessage: [{
                type: Input,
                args: ['fsFormCompareMessage']
            }] } });
