// @firestitch/form/esm2015/app/directives/validators/greater.directive.js
import { Directive, Input } from '@angular/core';
import { FsControlDirective } from './control.directive';
import { FsValidators } from '../../validators/validators';
import { VALIDATE_MESSAGE_PROVIDER } from '../../providers/validate-messages.provider';
import * as i0 from "@angular/core";
export class FsFormGreaterDirective extends FsControlDirective {
    set validationMessage(value) {
        this._validateMessages.greater = value;
    }
    ngOnChanges() {
        this._control.updateValueAndValidity();
    }
    validate(control) {
        const greater = parseFloat(this.fsFormGreater);
        const value = parseFloat(this._control.value);
        if (!isNaN(greater) && !isNaN(value) && value <= greater) {
            return { greater: { greater, actual: value } };
        }
        return FsValidators.numeric(this._control);
    }
}
FsFormGreaterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormGreaterDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
FsFormGreaterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsFormGreaterDirective, selector: "[fsFormGreater]", inputs: { fsFormGreater: "fsFormGreater", validationMessage: ["fsFormGreaterMessage", "validationMessage"] }, providers: [
        VALIDATE_MESSAGE_PROVIDER
    ], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormGreaterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsFormGreater]',
                    providers: [
                        VALIDATE_MESSAGE_PROVIDER
                    ],
                }]
        }], propDecorators: { fsFormGreater: [{
                type: Input
            }], validationMessage: [{
                type: Input,
                args: ['fsFormGreaterMessage']
            }] } });
