// @firestitch/form/esm2015/app/directives/validators/lesser.directive.js
import { Directive, Input } from '@angular/core';
import { FsControlDirective } from './control.directive';
import { FsValidators } from '../../validators/validators';
import { VALIDATE_MESSAGE_PROVIDER } from '../../providers/validate-messages.provider';
import * as i0 from "@angular/core";
export class FsFormLesserDirective extends FsControlDirective {
    set validationMessage(value) {
        this._validateMessages.lesser = value;
    }
    ngOnChanges() {
        this._control.updateValueAndValidity();
    }
    validate(control) {
        const lesser = parseFloat(this.fsFormLesser);
        const value = parseFloat(this._control.value);
        if (!isNaN(lesser) && !isNaN(value) && value >= lesser) {
            return { lesser: { lesser, actual: value } };
        }
        return FsValidators.numeric(this._control);
    }
}
FsFormLesserDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormLesserDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
FsFormLesserDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsFormLesserDirective, selector: "[fsFormLesser]", inputs: { fsFormLesser: "fsFormLesser", validationMessage: ["fsFormLesserMessage", "validationMessage"] }, providers: [
        VALIDATE_MESSAGE_PROVIDER
    ], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormLesserDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsFormLesser]',
                    providers: [
                        VALIDATE_MESSAGE_PROVIDER
                    ],
                }]
        }], propDecorators: { fsFormLesser: [{
                type: Input
            }], validationMessage: [{
                type: Input,
                args: ['fsFormLesserMessage']
            }] } });
