// @firestitch/form/esm2015/app/directives/validators/required.directive.js
import { Directive, Input } from '@angular/core';
import { Validators } from '@angular/forms';
import { FsControlDirective } from './control.directive';
import { VALIDATE_MESSAGE_PROVIDER } from '../../providers/validate-messages.provider';
import { isEnabled } from '../../helpers/is-enabled';
import * as i0 from "@angular/core";
export class FsFormRequiredDirective extends FsControlDirective {
    constructor() {
        super(...arguments);
        this.required = false;
    }
    set setFsFormRequired(value) {
        this.required = isEnabled(value);
    }
    set setRequired(value) {
        this.required = isEnabled(value);
    }
    set validationMessage(value) {
        this._validateMessages.required = value;
    }
    ngOnChanges() {
        this._control.updateValueAndValidity();
    }
    validate(control) {
        if (this.required) {
            return Validators.required(this._control);
        }
        else {
            return null;
        }
    }
    render() {
        const wrapper = this.getWrapperElement();
        if (wrapper && this.getlabelSelector()) {
            const labelWrapper = wrapper.querySelector(this.getlabelSelector());
            // Adding class fs-form-label-requried adds the * to the label
            if (labelWrapper) {
                if (this.required) {
                    this.renderer2.addClass(labelWrapper, 'fs-form-label-required');
                }
                else {
                    this.renderer2.removeClass(labelWrapper, 'fs-form-label-required');
                }
            }
        }
        super.render();
    }
}
FsFormRequiredDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormRequiredDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
FsFormRequiredDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsFormRequiredDirective, selector: "[fsFormRequired],[ngModel][required]", inputs: { setFsFormRequired: ["fsFormRequired", "setFsFormRequired"], setRequired: ["required", "setRequired"], validationMessage: ["fsFormRequiredMessage", "validationMessage"] }, providers: [
        VALIDATE_MESSAGE_PROVIDER
    ], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormRequiredDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsFormRequired],[ngModel][required]',
                    providers: [
                        VALIDATE_MESSAGE_PROVIDER
                    ],
                }]
        }], propDecorators: { setFsFormRequired: [{
                type: Input,
                args: ['fsFormRequired']
            }], setRequired: [{
                type: Input,
                args: ['required']
            }], validationMessage: [{
                type: Input,
                args: ['fsFormRequiredMessage']
            }] } });
