// @firestitch/datepicker/esm2015/libs/dialog/modules/desktop-dialog-container/components/header-month-range/header-month-range.component.js
import { ChangeDetectionStrategy, Component, Input, Output, EventEmitter, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/icon";
import * as i2 from "@firestitch/date";
export class FsDatePickerHeaderMonthRangeComponent {
    constructor() {
        this.goNextMonth = new EventEmitter();
        this.goPrevMonth = new EventEmitter();
    }
    get calendarMonth() {
        var _a;
        return (_a = this.calendarDate) === null || _a === void 0 ? void 0 : _a.getMonth();
    }
    get calendarYear() {
        var _a;
        return (_a = this.calendarDate) === null || _a === void 0 ? void 0 : _a.getFullYear();
    }
    get calendarDay() {
        var _a;
        return (_a = this.calendarDate) === null || _a === void 0 ? void 0 : _a.getDate();
    }
    nextMonth() {
        this.goNextMonth.emit();
    }
    prevMonth() {
        this.goPrevMonth.emit();
    }
}
FsDatePickerHeaderMonthRangeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerHeaderMonthRangeComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsDatePickerHeaderMonthRangeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDatePickerHeaderMonthRangeComponent, selector: "fs-date-picker-header-month-range", inputs: { calendarDate: "calendarDate", nextCalendarDate: "nextCalendarDate" }, outputs: { goNextMonth: "goNextMonth", goPrevMonth: "goPrevMonth" }, ngImport: i0, template: "<div class=\"current-month-header month-year\">\n  <div class=\"actions\">\n    <a (click)=\"prevMonth()\">\n      <mat-icon>navigate_before</mat-icon>\n    </a>\n  </div>\n  <span class=\"month-name\">\n    {{ calendarDate | fsDateFormat: 'full-date-dayless' }}\n  </span>\n</div>\n\n\n\n<div class=\"next-month-header month-year\">\n  <span class=\"month-name\">\n    {{ nextCalendarDate | fsDateFormat: 'full-date-dayless' }}\n  </span>\n  <div class=\"actions\">\n    <a (click)=\"nextMonth()\">\n      <mat-icon>navigate_next</mat-icon>\n    </a>\n  </div>\n</div>\n", styles: [":host{display:grid;grid-template-columns:1fr 1fr;grid-column-gap:40px}.month-year{padding:0;position:relative;justify-content:center}.current-month-header .actions{left:5px;position:absolute}.next-month-header .actions{right:5px;position:absolute}\n"], components: [{ type: i1.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }], pipes: { "fsDateFormat": i2.FsDateFormatPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerHeaderMonthRangeComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-date-picker-header-month-range',
                    templateUrl: './header-month-range.component.html',
                    styleUrls: [
                        './header-month-range.component.scss',
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { calendarDate: [{
                type: Input
            }], nextCalendarDate: [{
                type: Input
            }], goNextMonth: [{
                type: Output
            }], goPrevMonth: [{
                type: Output
            }] } });
