// @firestitch/form/esm2015/app/components/form-dialog-actions/form-dialog-actions.component.js
import { Component, ChangeDetectionStrategy, Input, Optional, ChangeDetectorRef } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';
import { Subject } from 'rxjs';
import { delay, filter, takeUntil } from 'rxjs/operators';
import { FsFormDirective } from '../../directives/form/form.directive';
import { ConfirmResult } from '../../enums/confirm-result';
import * as i0 from "@angular/core";
import * as i1 from "../../directives/form/form.directive";
import * as i2 from "@angular/material/dialog";
import * as i3 from "@angular/material/button";
import * as i4 from "@angular/common";
import * as i5 from "../../directives/submit-button.directive";
export class FsFormDialogActionsComponent {
    constructor(_form, _dialogRef, _cdRef) {
        this._form = _form;
        this._dialogRef = _dialogRef;
        this._cdRef = _cdRef;
        this.save = true;
        this.create = false;
        this.close = false;
        this.dirty = false;
        this._destroy$ = new Subject();
    }
    ngOnInit() {
        if (this._form) {
            this._form.ngForm.valueChanges
                .pipe(filter(() => (!this.dirty)), takeUntil(this._destroy$))
                .subscribe(() => {
                this.dirty = this._form.ngForm.dirty;
                this._cdRef.markForCheck();
            });
            this._form.submitted
                .pipe(delay(50), takeUntil(this._destroy$))
                .subscribe(() => {
                this.dirty = false;
                this._cdRef.markForCheck();
            });
            this._form.reseted
                .pipe(takeUntil(this._destroy$))
                .subscribe(() => {
                this.dirty = false;
                this._cdRef.markForCheck();
            });
        }
    }
    closeClick() {
        if (this._form) {
            this._form.triggerConfirm()
                .pipe(filter((confirmResult) => (confirmResult !== ConfirmResult.Review)), takeUntil(this._destroy$))
                .subscribe(() => {
                this._dialogRef.close(null);
            });
        }
        else {
            this._dialogRef.close(null);
        }
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
}
FsFormDialogActionsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormDialogActionsComponent, deps: [{ token: i1.FsFormDirective, optional: true }, { token: i2.MatDialogRef, optional: true }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
FsFormDialogActionsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsFormDialogActionsComponent, selector: "fs-form-dialog-actions", inputs: { save: "save", create: "create", close: "close", name: "name" }, ngImport: i0, template: "<div class=\"buttons\">\r\n  <ng-container *ngIf=\"save\">\r\n    <button \r\n        mat-button \r\n        type=\"submit\"\r\n        color=\"primary\" \r\n        [name]=\"name\">\r\n      {{create ? 'Create' : 'Save'}}\r\n    </button>\r\n    <button \r\n        mat-button \r\n        type=\"button\"\r\n        [disabled]=\"close && !dirty && !create\"\r\n        [matDialogClose]=\"null\">\r\n      Cancel\r\n    </button>\r\n  </ng-container>\r\n\r\n  <ng-content></ng-content>\r\n  <div class=\"close\" *ngIf=\"close\">\r\n    <button \r\n        mat-button \r\n        type=\"button\"\r\n        [color]=\"dirty ? 'basic' : 'primary'\"\r\n        (click)=\"closeClick()\">\r\n      Close\r\n    </button>\r\n  </div>\r\n</div>", styles: [".buttons{display:flex;flex-grow:1;align-items:center}.buttons .close{display:flex;flex-grow:1;justify-content:flex-end}:host{display:flex;flex-grow:1}@media only screen and (max-width: 599px){.buttons{flex-direction:column}}\n"], components: [{ type: i3.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }], directives: [{ type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i5.FsSubmitButtonDirective, selector: "button[type=\"submit\"]", inputs: ["name", "dirtySubmit"] }, { type: i2.MatDialogClose, selector: "[mat-dialog-close], [matDialogClose]", inputs: ["type", "mat-dialog-close", "aria-label", "matDialogClose"], exportAs: ["matDialogClose"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormDialogActionsComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-form-dialog-actions',
                    templateUrl: './form-dialog-actions.component.html',
                    styleUrls: ['./form-dialog-actions.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.FsFormDirective, decorators: [{
                    type: Optional
                }] }, { type: i2.MatDialogRef, decorators: [{
                    type: Optional
                }] }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { save: [{
                type: Input
            }], create: [{
                type: Input
            }], close: [{
                type: Input
            }], name: [{
                type: Input
            }] } });
