// @firestitch/clear/esm2015/app/components/clear-element/clear-element.component.js
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/icon";
import * as i2 from "@angular/common";
export class FsClearElementComponent {
    constructor(el) {
        this.el = el;
        this.clear = new EventEmitter();
        this.show = false;
    }
    ngAfterViewInit() {
        this.el.nativeElement.parentElement.parentElement.parentElement
            .appendChild(this.el.nativeElement.querySelector('.fs-clear-wrapper'));
    }
    clearClick(event) {
        this.clear.emit(event);
    }
}
FsClearElementComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearElementComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
FsClearElementComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsClearElementComponent, selector: "fs-clear", inputs: { show: "show" }, outputs: { clear: "clear" }, ngImport: i0, template: "\n\n<div class=\"mat-form-field-suffix fs-clear-wrapper\">\n  <mat-icon (click)=\"clearClick($event)\" *ngIf=\"show\">clear</mat-icon>\n</div>\n", styles: [".fs-clear-wrapper{align-self:flex-end}.fs-clear-wrapper mat-icon{cursor:pointer}\n"], components: [{ type: i1.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearElementComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-clear',
                    templateUrl: 'clear-element.component.html',
                    styleUrls: ['clear-element.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { clear: [{
                type: Output
            }], show: [{
                type: Input
            }] } });
