// @firestitch/form/esm2015/app/directives/form-dialog-close.directive.js
import { Directive, HostListener, HostBinding } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class FsFormDialogCloseDirective {
    constructor() {
        this.clicked$ = new Subject();
        this.registered = false;
        this.type = 'button';
    }
    click() {
        this.clicked$.next();
    }
    ngOnDestroy() {
        this.clicked$.complete();
    }
}
FsFormDialogCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormDialogCloseDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
FsFormDialogCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsFormDialogCloseDirective, selector: "[fsFormDialogClose],[fs-form-dialog-close]", host: { listeners: { "click": "click($event.target)" }, properties: { "attr.type": "this.type" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormDialogCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsFormDialogClose],[fs-form-dialog-close]'
                }]
        }], propDecorators: { type: [{
                type: HostBinding,
                args: ['attr.type']
            }], click: [{
                type: HostListener,
                args: ['click', ['$event.target']]
            }] } });
