// @firestitch/clear/esm2015/app/components/clear/clear.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../clear-element/clear-element.component";
export class FsClearComponent {
    constructor() {
        this.ngModel = null;
        this.ngModelChange = new EventEmitter();
        this.cleared = new EventEmitter();
    }
    clear() {
        this.ngModelChange.emit(null);
        this.cleared.emit();
    }
}
FsClearComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsClearComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsClearComponent, selector: "[fsClear]", inputs: { ngModel: "ngModel" }, outputs: { ngModelChange: "ngModelChange", cleared: "cleared" }, ngImport: i0, template: "<fs-clear (clear)=\"clear()\" [show]=\"ngModel\"></fs-clear>", styles: [""], components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsClear]',
                    templateUrl: 'clear.component.html',
                    styleUrls: ['clear.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { ngModel: [{
                type: Input
            }], ngModelChange: [{
                type: Output
            }], cleared: [{
                type: Output
            }] } });
