// @firestitch/datepicker/esm2015/libs/dialog/dialog.module.js
import { NgModule } from '@angular/core';
import { FsDatePickerDialogFactory } from './services/dialog-factory.service';
import { FsDatePickerDesktopDialogContainerModule } from './modules/desktop-dialog-container/desktop-dialog-container.module';
import { FsDatePickerScrollPickerDialogContainerModule } from './modules/scroll-picker-dialog-container/scroll-picker-dialog-container.module';
import { FsDatePickerMobileDialogContainerModule } from './modules/mobile-dialog-container/mobile-dialog-container.module';
import * as i0 from "@angular/core";
export class FsDatePickerDialogModule {
    static forRoot() {
        return {
            ngModule: FsDatePickerDialogModule,
            providers: [
                FsDatePickerDialogFactory,
            ]
        };
    }
}
FsDatePickerDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDatePickerDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDialogModule, imports: [FsDatePickerDesktopDialogContainerModule,
        FsDatePickerScrollPickerDialogContainerModule,
        FsDatePickerMobileDialogContainerModule] });
FsDatePickerDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDialogModule, imports: [[
            FsDatePickerDesktopDialogContainerModule,
            FsDatePickerScrollPickerDialogContainerModule,
            FsDatePickerMobileDialogContainerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        FsDatePickerDesktopDialogContainerModule,
                        FsDatePickerScrollPickerDialogContainerModule,
                        FsDatePickerMobileDialogContainerModule,
                    ],
                }]
        }] });
