// @firestitch/datepicker/esm2015/libs/calendar/models/period.js
import { addWeeks, differenceInCalendarYears, } from 'date-fns';
export class Period {
    constructor(periodId, startDate, seedDate, periodWeeks, _selected = false) {
        this.periodId = periodId;
        this.startDate = startDate;
        this.seedDate = seedDate;
        this.periodWeeks = periodWeeks;
        this._selected = _selected;
        this.mouseOver = false;
        this._weeks = [];
        this._countOfWeeks = 0;
        this._updatePeriodInterval();
    }
    get countOfWeeks() {
        return this._countOfWeeks;
    }
    get selected() {
        return this._selected;
    }
    get from() {
        return this._from;
    }
    get to() {
        return this._to;
    }
    get year() {
        return this._year;
    }
    set year(value) {
        this._year = value;
    }
    set selected(value) {
        this._selected = value;
        this._weeks.forEach((week) => {
            week.days.forEach((day) => {
                day.selected = value;
            });
        });
    }
    addWeek(week) {
        this._weeks.push(week);
        if (this.seedDate && this.periodWeeks) {
            this._sortWeeks();
            if (this._weeks.length === 1) {
                week.setPeriodVisibility(true);
            }
        }
        this._countOfWeeks = this._weeks.length;
    }
    /**
     * Mark weeks with special flags to be able to draw borders
     */
    markFirstLastWeeks() {
        const firstWeek = this._weeks[0];
        const lastWeek = this._weeks[this._weeks.length - 1];
        firstWeek.markAsFirstVisiblePeriodWeek();
        lastWeek.markAsLastVisiblePeriodWeek();
    }
    _sortWeeks() {
        this._weeks.sort((a, b) => {
            if (a.dateStart < b.dateStart) {
                return -1;
            }
            else if (a.dateStart > b.dateStart) {
                return 1;
            }
            else {
                return 0;
            }
        });
    }
    /**
     * Calc from, to, year params based on period date start
     */
    _updatePeriodInterval() {
        const diffInCalendarYears = differenceInCalendarYears(this.startDate, this.seedDate);
        const seedDate = new Date(this.seedDate);
        seedDate.setFullYear(seedDate.getFullYear() + diffInCalendarYears);
        const offsetFromSeedDate = (this.periodWeeks * (this.periodId - 1));
        this._from = addWeeks(seedDate, offsetFromSeedDate);
        this._to = addWeeks(this._from, this.periodWeeks);
        this._year = this._from.getFullYear();
    }
}
