// @firestitch/clear/esm2015/app/fs-clear.module.js
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { CommonModule } from '@angular/common';
import { FsClearComponent } from './components/clear/clear.component';
import { FsClearElementComponent } from './components/clear-element/clear-element.component';
import * as i0 from "@angular/core";
export class FsClearModule {
    static forRoot() {
        return {
            ngModule: FsClearModule
        };
    }
}
FsClearModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsClearModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearModule, declarations: [FsClearComponent,
        FsClearElementComponent], imports: [CommonModule,
        MatIconModule,
        MatButtonModule], exports: [FsClearComponent,
        FsClearElementComponent] });
FsClearModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearModule, imports: [[
            CommonModule,
            MatIconModule,
            MatButtonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsClearModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatIconModule,
                        MatButtonModule
                    ],
                    exports: [
                        FsClearComponent,
                        FsClearElementComponent
                    ],
                    declarations: [
                        FsClearComponent,
                        FsClearElementComponent
                    ]
                }]
        }] });
