// @firestitch/datepicker/esm2015/app/components/range-picker/from/date-range-picker-from.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { FsDatePickerComponent } from '../../date-picker/date-picker.component';
import { RangePickerFromComponent } from '../base/range-picker-from.component';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/clear";
import * as i2 from "../../date-picker-trigger/date-picker-trigger.component";
export class DateRangePickerFromComponent extends RangePickerFromComponent {
    set fsDateRangeFrom(value) {
        this._name = value;
    }
    set fsDateRangeFromPicker(value) {
        this._name = value;
    }
}
DateRangePickerFromComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangePickerFromComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
DateRangePickerFromComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: DateRangePickerFromComponent, selector: "[fsDateRangeFrom],[fsDateRangeFromPicker]", inputs: { fsDateRangeFrom: "fsDateRangeFrom", fsDateRangeFromPicker: "fsDateRangeFromPicker" }, usesInheritance: true, ngImport: i0, template: "\n    <fs-clear [show]=\"value && !disabled && !readonly && clear\" (clear)=\"cleared($event)\"></fs-clear>\n    <fs-datepicker-trigger (click)=\"triggerClick()\" [disabled]=\"disabled || readonly\" [view]=\"view\"></fs-datepicker-trigger>\n  ", isInline: true, components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }, { type: i2.FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: ["disabled", "view"], outputs: ["click"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateRangePickerFromComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsDateRangeFrom],[fsDateRangeFromPicker]',
                    template: FsDatePickerComponent.template,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { fsDateRangeFrom: [{
                type: Input
            }], fsDateRangeFromPicker: [{
                type: Input
            }] } });
