// @firestitch/datepicker/esm2015/app/components/range-picker/from/time-range-picker-from.component.js
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
import { PickerViewType } from '../../../../libs/common/enums/picker-view-type.enum';
import { FsDatePickerComponent } from '../../date-picker/date-picker.component';
import { RangePickerFromComponent } from '../base/range-picker-from.component';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/clear";
import * as i2 from "../../date-picker-trigger/date-picker-trigger.component";
export class TimeRangePickerFromComponent extends RangePickerFromComponent {
    constructor() {
        super(...arguments);
        this.view = PickerViewType.Time;
    }
    set fsTimeRangeFrom(value) {
        this._name = value;
    }
    set fsTimeRangeFromPicker(value) {
        this._name = value;
    }
    updateValue(value) {
        this._pickerRef.updateStartDate(value);
        super.updateValue(value);
    }
}
TimeRangePickerFromComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TimeRangePickerFromComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TimeRangePickerFromComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: TimeRangePickerFromComponent, selector: "[fsTimeRangeFrom],[fsTimeRangeFromPicker]", inputs: { fsTimeRangeFrom: "fsTimeRangeFrom", fsTimeRangeFromPicker: "fsTimeRangeFromPicker", view: "view" }, usesInheritance: true, ngImport: i0, template: "\n    <fs-clear [show]=\"value && !disabled && !readonly && clear\" (clear)=\"cleared($event)\"></fs-clear>\n    <fs-datepicker-trigger (click)=\"triggerClick()\" [disabled]=\"disabled || readonly\" [view]=\"view\"></fs-datepicker-trigger>\n  ", isInline: true, components: [{ type: i1.FsClearElementComponent, selector: "fs-clear", inputs: ["show"], outputs: ["clear"] }, { type: i2.FsDatePickerTriggerComponent, selector: "fs-datepicker-trigger", inputs: ["disabled", "view"], outputs: ["click"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TimeRangePickerFromComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsTimeRangeFrom],[fsTimeRangeFromPicker]',
                    template: FsDatePickerComponent.template,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { fsTimeRangeFrom: [{
                type: Input
            }], fsTimeRangeFromPicker: [{
                type: Input
            }], view: [{
                type: Input
            }] } });
