// @firestitch/datepicker/esm2015/libs/dialog/modules/scroll-picker-dialog-container/components/date-scroll-picker-desktop/date-scroll-picker-desktop.component.js
import { ChangeDetectionStrategy, Component, } from '@angular/core';
import { FsDatePickerDialogRef } from '../../../../../dialog/classes/dialog-ref';
import * as i0 from "@angular/core";
import * as i1 from "../../../../../dialog/classes/dialog-ref";
import * as i2 from "../date-scroll-picker/date-scroll-picker.component";
import * as i3 from "@angular/material/button";
export class FsDateScrollPickerDesktopComponent {
    constructor(_dialogRef) {
        this._dialogRef = _dialogRef;
    }
    get dialogRef() {
        return this._dialogRef;
    }
    change(date) {
        this._date = date;
    }
    close(save = false) {
        if (save) {
            this._dialogRef.updateValue(this._date);
        }
        this._dialogRef.close();
    }
}
FsDateScrollPickerDesktopComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateScrollPickerDesktopComponent, deps: [{ token: i1.FsDatePickerDialogRef }], target: i0.ɵɵFactoryTarget.Component });
FsDateScrollPickerDesktopComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDateScrollPickerDesktopComponent, selector: "ng-component", ngImport: i0, template: "<div class=\"fs-date-picker-dialog fs-date-picker-dialog-shadow fs-date-picker-range\" tabindex=\"0\">\n  <div class=\"wrap\">\n    <fs-date-scroll-picker [model]=\"dialogRef.pickerModel.model\"\n                           [minYear]=\"dialogRef.options.minYear\"\n                           [maxDate]=\"dialogRef.options.maxDate\"\n                           [maxYear]=\"dialogRef.options.maxYear\"\n                           [showDay]=\"dialogRef.options.showDay\"\n                           [showMonth]=\"dialogRef.options.showMonth\"\n                           [showYear]=\"dialogRef.options.showYear\"\n                           (changed)=\"change($event)\">\n    </fs-date-scroll-picker>\n\n    <div class=\"buttons\">\n      <button type=\"button\" mat-button color=\"primary\" (click)=\"close(true)\">Select</button>\n      <button type=\"button\" mat-button (click)=\"close(false)\">Cancel</button>\n    </div>\n  </div>\n</div>\n", styles: [""], components: [{ type: i2.FsDateScrollPickerDialogComponent, selector: "fs-date-scroll-picker", inputs: ["model", "showMonth", "showDay", "showYear", "minYear", "maxYear", "maxDate"], outputs: ["changed"] }, { type: i3.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDateScrollPickerDesktopComponent, decorators: [{
            type: Component,
            args: [{
                    templateUrl: './date-scroll-picker-desktop.component.html',
                    styleUrls: ['./date-scroll-picker-desktop.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.FsDatePickerDialogRef }]; } });
