// @firestitch/datepicker/esm2015/app/services/range-picker-store.service.js
import { Injectable } from '@angular/core';
import { RangePickerRef } from '../classes/range-picker-ref';
import * as i0 from "@angular/core";
export class FsRangePickerStoreService {
    constructor() {
        this._store = new Map();
    }
    registerPickerFrom(name, value, view) {
        this._createIfNotExistsPicker(name, view);
        const pickerRef = this._store.get(name);
        // pickerRef.updateStartDate(value);
        return pickerRef;
    }
    destroyStartDatePicker(name) {
        const pickerRef = this._store.get(name);
        if (pickerRef) {
            pickerRef.destroyStartDatePicker();
            this._destroyRefIfRelatedDestroyed(name, pickerRef);
        }
    }
    destroyEndDatePicker(name) {
        const pickerRef = this._store.get(name);
        if (pickerRef) {
            pickerRef.destroyEndDatePicker();
            this._destroyRefIfRelatedDestroyed(name, pickerRef);
        }
    }
    registerPickerTo(name, value, view) {
        this._createIfNotExistsPicker(name, view);
        const pickerRef = this._store.get(name);
        // pickerRef.updateEndDate(value);
        return pickerRef;
    }
    _createIfNotExistsPicker(name, view) {
        if (!this._store.has(name)) {
            this._store.set(name, new RangePickerRef(view));
        }
    }
    _destroyRefIfRelatedDestroyed(name, ref) {
        if (!ref.startDatePickerExists && !ref.endDatePickerExists) {
            ref.destroy();
            this._store.delete(name);
        }
    }
}
FsRangePickerStoreService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsRangePickerStoreService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
FsRangePickerStoreService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsRangePickerStoreService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsRangePickerStoreService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
