// @firestitch/datepicker/esm2015/libs/dialog/modules/mobile-dialog-container/mobile-dialog-container.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatTabsModule } from '@angular/material/tabs';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatListModule } from '@angular/material/list';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { FsDateModule } from '@firestitch/date';
import { FsDatePickerCalendarModule } from '../../../calendar/calendar.module';
import { FsMobileCalendarDialogComponent } from './components/mobile-dialog/mobile-dialog.component';
import { FsDatePickerMobileHeaderComponent } from './components/header/header.component';
import { FsMobileTimePickerComponent } from './components/time-picker/time-picker.component';
import { FDatePickerMobileDatetimePickerComponent } from './components/date-time-picker/date-time-picker.component';
import { FsDatePickerMobileHeaderMonthRangeComponent } from './components/header-month-range/header-month-range.component';
import { FsDatePickerVirtualScrollCalendarComponent } from './components/virtual-scroll-calendar/virtual-scroll-calendar.component';
import * as i0 from "@angular/core";
export class FsDatePickerMobileDialogContainerModule {
}
FsDatePickerMobileDialogContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileDialogContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsDatePickerMobileDialogContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileDialogContainerModule, declarations: [FsMobileCalendarDialogComponent,
        FsDatePickerMobileHeaderComponent,
        FsMobileTimePickerComponent,
        FDatePickerMobileDatetimePickerComponent,
        FsDatePickerMobileHeaderMonthRangeComponent,
        FsDatePickerVirtualScrollCalendarComponent], imports: [CommonModule,
        MatTabsModule,
        MatButtonModule,
        MatIconModule,
        FsDatePickerCalendarModule,
        FsDateModule,
        ScrollingModule,
        MatListModule], exports: [FsMobileCalendarDialogComponent] });
FsDatePickerMobileDialogContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileDialogContainerModule, imports: [[
            CommonModule,
            MatTabsModule,
            MatButtonModule,
            MatIconModule,
            FsDatePickerCalendarModule,
            FsDateModule,
            ScrollingModule,
            MatListModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDatePickerMobileDialogContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatTabsModule,
                        MatButtonModule,
                        MatIconModule,
                        FsDatePickerCalendarModule,
                        FsDateModule,
                        ScrollingModule,
                        MatListModule,
                    ],
                    declarations: [
                        FsMobileCalendarDialogComponent,
                        FsDatePickerMobileHeaderComponent,
                        FsMobileTimePickerComponent,
                        FDatePickerMobileDatetimePickerComponent,
                        FsDatePickerMobileHeaderMonthRangeComponent,
                        FsDatePickerVirtualScrollCalendarComponent,
                    ],
                    exports: [
                        FsMobileCalendarDialogComponent,
                    ],
                }]
        }] });
