// @firestitch/autocomplete/esm2015/app/fs-autocomplete.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatChipsModule } from '@angular/material/chips';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { FsAutocompleteComponent } from './components/autocomplete/autocomplete.component';
import { FsAutocompleteTemplateDirective } from './directives/autocomplete-template/autocomplete-template.directive';
import { FsAutocompleteSuffixDirective } from './directives/autocomplete-suffix/autocomplete-suffix.directive';
import { FsAutocompleteStaticDirective } from './directives/autocomplete-static/autocomplete-static.directive';
import { FsAutocompleteNoResultsDirective } from './directives/no-results-template/no-results-template.directive';
import { FsAutocompleteHintDirective } from './directives/autocomplete-hint/autocomplete-hint.directive';
import * as i0 from "@angular/core";
export class FsAutocompleteModule {
}
FsAutocompleteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsAutocompleteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteModule, declarations: [FsAutocompleteComponent,
        FsAutocompleteTemplateDirective,
        FsAutocompleteSuffixDirective,
        FsAutocompleteStaticDirective,
        FsAutocompleteNoResultsDirective,
        FsAutocompleteHintDirective], imports: [CommonModule,
        FormsModule,
        MatAutocompleteModule,
        MatChipsModule,
        MatInputModule,
        MatIconModule], exports: [FsAutocompleteComponent,
        FsAutocompleteTemplateDirective,
        FsAutocompleteSuffixDirective,
        FsAutocompleteStaticDirective,
        FsAutocompleteNoResultsDirective,
        FsAutocompleteHintDirective] });
FsAutocompleteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteModule, providers: [], imports: [[
            CommonModule,
            FormsModule,
            MatAutocompleteModule,
            MatChipsModule,
            MatInputModule,
            MatIconModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsAutocompleteModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        MatAutocompleteModule,
                        MatChipsModule,
                        MatInputModule,
                        MatIconModule,
                    ],
                    exports: [
                        FsAutocompleteComponent,
                        FsAutocompleteTemplateDirective,
                        FsAutocompleteSuffixDirective,
                        FsAutocompleteStaticDirective,
                        FsAutocompleteNoResultsDirective,
                        FsAutocompleteHintDirective,
                    ],
                    declarations: [
                        FsAutocompleteComponent,
                        FsAutocompleteTemplateDirective,
                        FsAutocompleteSuffixDirective,
                        FsAutocompleteStaticDirective,
                        FsAutocompleteNoResultsDirective,
                        FsAutocompleteHintDirective,
                    ],
                    providers: [],
                }]
        }] });
